Unclear error message with mutually recursive methods with inferred return types

```scala
class A {
  def odd(x: Int) = if (x == 0) false else !even(x-1)
  def even(x: Int) = if (x == 0) true else !odd(x-1)
}
```
dotty:
```scala
3 |  def even(x: Int) = if (x == 0) true else !odd(x-1)
  |                                            ^^^
  |                                     cyclic reference involving method odd
```
Contrast with scalac:
```scala
mutrec.scala:3: error: recursive method odd needs result type
  def even(x: Int) = if (x == 0) true else !odd(x-1)
                                            ^
```

--- a/compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala
@@ -40,12 +40,12 @@ object ErrorReporting {
         }
       } else msg
 
-      if (cycleSym.is(Implicit, butNot = Method) && cycleSym.owner.isTerm)
-        em"""cyclic reference involving implicit $cycleSym
+    if (cycleSym.is(Implicit, butNot = Method) && cycleSym.owner.isTerm)
+      em"""cyclic reference involving implicit $cycleSym
             |This happens when the right hand-side of $cycleSym's definition involves an implicit search.
             |To avoid the error, give $cycleSym an explicit type."""
-      else
-        errorMsg(ex.show, ctx)
+    else
+      errorMsg(ex.show, ctx)
   }
 
   def wrongNumberOfTypeArgs(fntpe: Type, expectedArgs: List[TypeParamInfo], actual: List[untpd.Tree], pos: Position)(implicit ctx: Context) =

--- a/compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala
@@ -31,17 +31,10 @@ object ErrorReporting {
     def errorMsg(msg: String, cx: Context): String =
       if (cx.mode is Mode.InferringReturnType) {
         cx.tree match {
-          case tree: untpd.ValOrDefDef =>
-              // Dotty deviation: Was Trees.ValOrDefDef[_], but this gives ValOrDefDef[Nothing] instead of
-              // ValOrDefDel[Null]. Scala handles it, but it looks accidental because bounds propagation
-              // fails if the parameter is invariant or cotravariant.
-              // See test pending/pos/boundspropagation.scala
-            val treeSym = ctx.symOfContextTree(tree)
-            if (treeSym.exists && treeSym.name == cycleSym.name && treeSym.owner == cycleSym.owner) {
-              val result = if (cycleSym is Method) " result" else ""
-              em"overloaded or recursive $cycleSym needs$result type"
-            }
-            else errorMsg(msg, cx.outer)
+          case tree: untpd.DefDef if !tree.tpt.typeOpt.exists =>
+            em"overloaded or recursive method ${tree.name} needs result type"
+          case tree: untpd.ValDef if !tree.tpt.typeOpt.exists =>
+            em"recursive value ${tree.name} needs type"
           case _ =>
             errorMsg(msg, cx.outer)
         }

--- a/tests/neg/i2001.scala
+++ b/tests/neg/i2001.scala
@@ -1,0 +1,6 @@
+class A {
+  def odd(x: Int) = if (x == 0) false else !even(x-1)
+  def even(x: Int) = if (x == 0) true else !odd(x-1) // error: overloaded or recursive method needs result type
+
+  lazy val x = x // error: recursive value needs type
+}

